dpnp/backend/extensions/indexing/choose.cpp

```diff
@@ -0,0 +1,455 @@
+    if (overlap(src, dst)) {
+        throw py::value_error("Array memory overlap.");
+    }
+
+    py::ssize_t src_offset = py::ssize_t(0);
+    py::ssize_t dst_offset = py::ssize_t(0);
+
+    int src_typenum = src.get_typenum();
+    int dst_typenum = dst.get_typenum();
+    int chc_typenum = chc_rep.get_typenum();
+
+    auto array_types = td_ns::usm_ndarray_types();
+    int src_type_id = array_types.typenum_to_lookup_id(src_typenum);
+    int dst_type_id = array_types.typenum_to_lookup_id(dst_typenum);
+    int chc_type_id = array_types.typenum_to_lookup_id(chc_typenum);
+
+    if (chc_type_id != dst_type_id) {
+        throw py::type_error("Output and choice data types are not the same.");
+    }
+
+    dpctl::tensor::validation::AmpleMemory::throw_if_not_ample(dst, nelems);
+
+    std::vector<char *> chc_ptrs;
+    chc_ptrs.reserve(n_chcs);
+
+    std::vector<py::ssize_t> chc_offsets;
+    chc_offsets.reserve(n_chcs);
+
+    auto sh_nelems = std::max<int>(nd, 1);
+    std::vector<py::ssize_t> chc_strides(n_chcs * sh_nelems, 0);
+
+    // first iteration for first choice array chc_rep
+    if (overlap(dst, chc_rep)) {
+        throw py::value_error("Arrays index overlapping segments of memory");
+    }
+
+    // chc_strides is initialized to 0 for 0D choices, so skip
+    if (nd > 0) {
+        auto chc_strides_ = chc_rep.get_strides_vector();
+        std::copy(chc_strides_.begin(), chc_strides_.end(),
+                  chc_strides.begin());
+    }
+
+    chc_ptrs.push_back(chc_rep.get_data());
+    chc_offsets.push_back(py::ssize_t(0));
+
+    for (auto i = 1; i < n_chcs; ++i) {
+        dpctl::tensor::usm_ndarray chc_ = chcs[i];
+
+        // ndim, type, and shape are checked against the first array
+        if (!(chc_.get_ndim() == nd)) {
+            throw py::value_error("Choice array dimensions are not the same");
+        }
+
+        if (!(chc_type_id ==
+              array_types.typenum_to_lookup_id(chc_.get_typenum()))) {
+            throw py::type_error(
+                "Choice array data types are not all the same.");
+        }
+
+        const py::ssize_t *chc_shape_ = chc_.get_shape_raw();
+        for (int dim = 0; dim < nd; ++dim) {
+            if (!(chc_shape[dim] == chc_shape_[dim])) {
+                throw py::value_error("Choice shapes are not all equal.");
+            }
+        }
+
+        // check for overlap with destination
+        if (overlap(dst, chc_)) {
+            throw py::value_error(
+                "Arrays index overlapping segments of memory");
+        }
+
+        char *chc_data = chc_.get_data();
+
+        if (nd > 0) {
+            auto chc_strides_ = chc_.get_strides_vector();
+            std::copy(chc_strides_.begin(), chc_strides_.end(),
+                      chc_strides.begin() + i * nd);
+        }
+
+        chc_ptrs.push_back(chc_data);
+        chc_offsets.push_back(py::ssize_t(0));
+    }
+
+    char **packed_chc_ptrs = sycl::malloc_device<char *>(n_chcs, exec_q);
```

Comment:
`make_unique` cannot be used this way it seems, but something like
```
    auto usm_deleter = [q](char **usm){ sycl::free(usm, q); };
    using usm_unique_ptr = std::unique_ptr<char *, decltype(usm_deleter)>;
    auto chcs = usm_unique_ptr(sycl::malloc_device<char *>(n_chcs, q), usm_deleter);
```
does, and can probably be generalized to do the trick.

Unfortunately, `cgh.host_task([pcp = std::move(packed_chc_ptrs), ctx]() { });` also does not compile, but something like
`cgh.host_task([&chcs, ctx]() { sycl_free_noexcept(chcs.get(), ctx); });` does

So this can be used to avoid manually deallocating after every possible `runtime_error`, but the memory must still be freed manually in the host task.

Reply:
This example works for me:
```C++
#include <iostream>
#include <string>
#include <utility>
#include <memory>

int main() {
    auto p = std::make_unique<std::string>("Hello, world!");

    auto lambda = [p_ = std::move(p)]() {
        std::cout << *p_ << std::endl;
    };

    lambda(); // Output: Hello, world!
}
```

```
icpx -std=c++17 lambda_move.cpp
./a.out
```
Output:
```
Hello, world!
```

dpnp compiles with C++17 standard, so it should work

Reply:
Also capturing `unique_ptr` in lambda is terrible idea. `unique_ptr` would be destroyed as soon as you leave function.

Reply:
> This example works for me:
> 
> ```c++
> #include <iostream>
> #include <string>
> #include <utility>
> #include <memory>
> 
> int main() {
>     auto p = std::make_unique<std::string>("Hello, world!");
> 
>     auto lambda = [p_ = std::move(p)]() {
>         std::cout << *p_ << std::endl;
>     };
> 
>     lambda(); // Output: Hello, world!
> }
> ```
> 
> ```
> icpx -std=c++17 lambda_move.cpp
> ./a.out
> ```
> 
> Output:
> 
> ```
> Hello, world!
> ```
> 
> dpnp compiles with C++17 standard, so it should work



> This example works for me:
> 
> ```c++
> #include <iostream>
> #include <string>
> #include <utility>
> #include <memory>
> 
> int main() {
>     auto p = std::make_unique<std::string>("Hello, world!");
> 
>     auto lambda = [p_ = std::move(p)]() {
>         std::cout << *p_ << std::endl;
>     };
> 
>     lambda(); // Output: Hello, world!
> }
> ```
> 
> ```
> icpx -std=c++17 lambda_move.cpp
> ./a.out
> ```
> 
> Output:
> 
> ```
> Hello, world!
> ```
> 
> dpnp compiles with C++17 standard, so it should work

This example works fine, but the following:
```
#include <sycl/sycl.hpp>
#include <memory>
#include <cstddef>

int main() {
    sycl::queue q{};
    std::size_t n_chcs = 10;

    auto usm_deleter = [q](char **usm){ sycl::free(usm, q); };
    auto chcs = std::unique_ptr<char *, decltype(usm_deleter)>(sycl::malloc_device<char *>(n_chcs, q), usm_deleter);

    sycl::event ev = q.submit([&](sycl::handler &cgh) {
        const auto &ctx = q.get_context();
        cgh.host_task([pcp = std::move(chcs), ctx]() { });
    });

    return 0;
}
```
does not compile, as `chcs` is not copy-constructible, and therefore cannot be converted to `std::function`, which happens in the `host_task`.

Reply:
> Also capturing `unique_ptr` in lambda is terrible idea. `unique_ptr` would be destroyed as soon as you leave function.
> 
> you can do instead:
> 
> ```c++
> auto chcs_ = chcs.get();
> chcs.reset();
> cgh.host_task([chcs_, ctx]() { sycl_free_noexcept(chcs_, ctx); });
> ```
> 
> but moving is better option

Yes, good point, I hadn't thought that example through. I'll go ahead and make the change with `unique_ptr`

Reply:
> does not compile, as chcs is not copy-constructible, and therefore cannot be converted to std::function, which happens in the host_task.

Thanks for clarification.

Btw, it is better to use `release` instead of `reset`:
```c++
auto chcs_ = chcs.release();
cgh.host_task([chcs_, ctx]() { sycl_free_noexcept(chcs_, ctx); });
```

Also you can use some trivial wrapper like this for unique_ptr creation:
```c++
template<typename T, typename D>
auto as_unique_ptr(T *ptr, D&& deleter) {
    return std::unique_ptr<T, D>(ptr, std::forward<D>(deleter));
}

...
    auto p = as_unique_ptr(new std::string("Hello, world!"), [](std::string *p) {
        delete p;
    });
...
```

Reply:
This is probably more correct version of wrapper:
```c++
template<typename T, typename D>
auto as_unique_ptr(T *ptr, D&& deleter) {
    return std::unique_ptr<T, std::remove_reference_t<D>>(ptr, std::forward<D>(deleter));
}
```

Reply:
I've added in something similar, which takes a size of allocation and returns a unique_ptr with the deleter, to avoid tedious repetition writing the deleters (all identical)

I think something like this may be implemented in dpctl as well, great suggestion!